src/clr-angular/forms/input/input-container.spec.ts

```diff
@@ -0,0 +1,107 @@
+
+        beforeEach(() => {
+            TestBed.configureTestingModule({
+                imports: [ClrIconModule, ClrCommonFormsModule, FormsModule],
+                declarations: [ClrInputContainer, ClrInput, SimpleTest],
+                providers: [NgControl]
+            });
+            fixture = TestBed.createComponent(SimpleTest);
+            fixture.detectChanges();
+
+            inputContainerDE = fixture.debugElement.query(By.directive(ClrInputContainer));
+            inputContainer = inputContainerDE.componentInstance;
+            inputContainerEl = inputContainerDE.nativeElement;
+        });
+
+        it("injects the ControlService and subscribes in AfterContentInit", () => {
+            expect(inputContainerDE.injector.get(ControlService, null)).toBeTruthy();
+            expect(inputContainer.subscription).toBeTruthy();
+        });
+
+        it("projects the label", () => {
+            const label = inputContainerEl.querySelector("label");
+            expect(label).toBeTruthy();
+            expect(label.previousElementSibling).toBeFalsy();
+        });
+
+        it("projects the input", () => {
+            expect(inputContainerEl.querySelector("[clrInput]")).toBeTruthy();
+        });
+
+        it("projects the helper text", () => {
+            expect(inputContainerEl.querySelector("clr-control-helper")).toBeTruthy();
+        });
+
+        it("sets error classes and displays the icon when invalid", () => {
+            expect(inputContainerEl.querySelector(".clr-control-container").classList.contains("clr-error"))
+                .toBeFalse();
+            expect(inputContainerEl.querySelector(".clr-validate-icon")).toBeFalsy();
+            inputContainer.invalid = true;
+            fixture.detectChanges();
+            expect(inputContainerEl.querySelector(".clr-control-container").classList.contains("clr-error")).toBeTrue();
+            expect(inputContainerEl.querySelector(".clr-validate-icon")).toBeTruthy();
+        });
+
+        it("projects the error helper when invalid", () => {
+            expect(inputContainerEl.querySelector("clr-control-error")).toBeFalsy();
+            inputContainer.invalid = true;
+            fixture.detectChanges();
+            expect(inputContainerEl.querySelector("clr-control-error")).toBeTruthy();
+        });
+
+        it("adds the .clr-form-control class to the host", () => {
+            expect(inputContainerEl.classList).toContain("clr-form-control");
+        });
+
+        it("tracks the validity of the form control", () => {
+            const input = fixture.debugElement.query(By.directive(ClrInput)).nativeElement;
+            expect(inputContainer.invalid).toBeFalse();
+            // inputs must be both invalid and blurred to register the validity
+            input.value = "abc";
+            input.dispatchEvent(new Event("input"));
+            input.dispatchEvent(new Event("blur"));
+            fixture.detectChanges();
```

Comment:
You're re-testing the blur binding of the input here. You should directly set the value of the FormControl (and make it dirty if needed) rather than "type" into the input.